src/utils/utxo.ts

```diff
@@ -0,0 +1,159 @@
+    default:
+      throw new Error(`Unknown algorithm ${algorithm}`);
+  }
+}
+
+/**
+ * Select utxos to fill the amount required.
+ * This method should be faster since it stops the iteration once the target amount is reached.
+ * Obs: Does not work with authority utxos.
+ *
+ * @param {IStorage} storage The wallet storage to select the utxos
+ * @param {string} token The token uid to select the utxos
+ * @param {number} amount The target amount of tokens required
+ * @returns {Promise<{ utxos: IUtxo[], utxosAmount: number}>
+ */
+export async function fastUtxoSelection(
+  storage: IStorage,
+  token: string,
+  amount: number,
+): Promise<{ utxos: IUtxo[], amount: number}> {
+  const utxos: IUtxo[] = [];
+  let utxosAmount = 0;
+
+  const options: IUtxoFilterOptions = {
+    token,
+    authorities: 0,
+    target_amount: amount,
+    only_available_utxos: true,
+    order_by_value: 'desc',
+  };
+
+  for await (const utxo of storage.selectUtxos(options)) {
+    // We let selectUtxos to filter the utxos for us and stop after target amount is reached
+    utxosAmount += utxo.value;
+    utxos.push(utxo);
+  }
+
+  if (utxosAmount < amount) {
+    // Not enough funds to fill the amount required.
+    return {
+      utxos: [],
+      amount: 0,
+    };
+  }
+
+  return { utxos, amount: utxosAmount };
+}
+
+/**
+ * Select utxos to fill the amount required.
+ * This method will select the smallest utxos that are bigger than the amount required.
+ * Obs: this will iterate on all available utxos to choose the best suited selection.
+ * Obs: Does not work with authority utxos.
+ *
+ * @param {IStorage} storage The wallet storage to select the utxos
+ * @param {string} token The token uid to select the utxos
+ * @param {number} amount The target amount of tokens required
+ * @returns {Promise<{ utxos: IUtxo[], utxosAmount: number}>
+ */
+export async function bestUtxoSelection(
+  storage: IStorage,
+  token: string,
+  amount: number,
+): Promise<{ utxos: IUtxo[], amount: number}> {
+  const utxos: IUtxo[] = [];
+  let utxosAmount = 0;
+  let selectedUtxo: IUtxo|null = null;
+
+  const options: IUtxoFilterOptions = {
+    token,
+    authorities: 0,
+    only_available_utxos: true,
+    order_by_value: 'desc',
+  };
+  for await (const utxo of storage.selectUtxos(options)) {
+    // storage ensures the utxo can be used
+    if (utxo.value === amount) {
+      return {
+        utxos: [utxo],
+        amount: amount,
+      };
+    }
+
+    utxos.push(utxo);
+    utxosAmount += utxo.value;
+
+    if (utxo.value > amount) {
+      // We want to select the smallest utxo that is bigger than the amount
+      if (selectedUtxo === null || utxo.value < selectedUtxo.value) {
+        selectedUtxo = utxo;
+      }
+    }
+
+    if ((utxo.value < amount) && (selectedUtxo !== null)) {
+      // We already have an utxo that is bigger than the amount required
+      // with the lowest possible value.
+      // We don't need to iterate more
+      break;
```

Comment:
Why don't you return here directly?

```
  return {
    utxos: [selectedUtxo],
    amount: selectedUtxo.value,
  };
```

Reply:
It will be returned just after the loop, so I thought we could just break instead of returning the selectedUtxo in 2 places